src/types.py

```diff
@@ -599,164 +846,106 @@ def unify_with_args(self, concrete: "FunctionSignature") -> bool:
         return can_unify
 
 
-def type_from_ctype(ctype: CType, typemap: TypeMap, array_decay: bool = True) -> Type:
-    real_ctype = resolve_typedefs(ctype, typemap)
-    if isinstance(real_ctype, ca.ArrayDecl):
-        inner_type, dim = array_type_and_dim(real_ctype, typemap)
-        if array_decay:
-            return Type.ptr(inner_type)
-        size = inner_type.get_size_bits()
-        if size is not None and dim is not None:
-            size *= dim
-        else:
-            size = None
-        return Type._ctype(real_ctype, typemap, size=size)
-    if isinstance(real_ctype, ca.PtrDecl):
-        return Type.ptr(type_from_ctype(real_ctype.type, typemap, array_decay=False))
-    if isinstance(real_ctype, ca.FuncDecl):
-        fn = parse_function(real_ctype)
-        assert fn is not None
-        fn_sig = FunctionSignature(
-            return_type=Type.void(),
-            is_variadic=fn.is_variadic,
-        )
-        if fn.ret_type is not None:
-            fn_sig.return_type = type_from_ctype(fn.ret_type, typemap)
-        if fn.params is not None:
-            fn_sig.params = [
-                FunctionParam(
-                    name=param.name or "",
-                    type=type_from_ctype(param.type, typemap),
-                )
-                for param in fn.params
-            ]
-            fn_sig.params_known = True
-        return Type.function(fn_sig)
-    if isinstance(real_ctype, ca.TypeDecl):
-        if isinstance(real_ctype.type, (ca.Struct, ca.Union)):
-            struct = parse_struct(real_ctype.type, typemap)
-            return Type._ctype(struct.type, typemap, size=struct.size * 8)
-        names = (
-            ["int"] if isinstance(real_ctype.type, ca.Enum) else real_ctype.type.names
+@dataclass(eq=False)
+class StructDeclaration:
+    """Representation of a C struct or union"""
+
+    @dataclass(eq=False)
+    class StructField:
+        type: Type
+        offset_bits: int
+        name: str
+
+    size_bits: int
+    align_bits: int
```

Comment:
I think bytes would be better, since structs are always byte-aligned. For fields it does make more sense to use bit measurements. Although... I suspect if we introduce bitfields we might want those to live entirely within StructDeclaration and not use Type's with size not divisible by 8? The Type would instead mirror the bitfield's underlying type, or just be "int". That way we prevent those weird types from leaking and accidentally ending up behind pointers etc., and bitfield accesses go through special paths that don't benefit from a uniform representation anyway. I think it's fair to say that we will never auto-detect bitfields, which simplifies the story a bit. Given this train of thought maybe all our usage of bit-based sizes should go away...

Reply:
This is interesting. I was not feeling particularly excited about supporting bitfields, but I was hoping to not make it *harder* :smile: 

I didn't think too much about it, but I was approximately guessing that bitfields could be represented as say, `Type.int_of_size(3)`, like you mentioned.
I hadn't thought about how these types might leak out though -- that's an excellent point.

I think you made a good argument for using bytes instead, I'll try to make that change.

Reply:
Yeah, `Type.int_of_size(3)` was my thinking too up until I wrote that comment. I'm also not too excited about bitfields, they are a bit too rare and complex to be worth the trouble

Reply:
Changed to bytes; renamed fields :+1: